### PR TITLE
[GraphScheduler] Add a simple topological sort graph scheduler

### PR DESCRIPTION
**Pull request by QiJune**
*Description*: Add a simple topological sort graph scheduler
*Testing*:
*Documentation*:

There are many scheduling policies to mapping a computation graph to a certain backend. This PR adds a simple topological sort based graph scheduler and exposes a command line interface for users.

### CODE REVIEW

**Comment by rdzhabarov** [comment]
@QiJune can you explain the motivation for this change?

IIRC, we had a similar scheduling schema originally and @opti-mix made that more efficient.

**Comment by QiJune** [comment]
@rdzhabarov  Actually, I want to implement a graph scheduler for a multi-core processor backend. And here is the design [doc](https://docs.google.com/document/d/1mC3KR76ljTN7bZ_712ofAJjcy1nKxmzKD6GPyXnM1nk/edit). This PR is just a trial to add a new scheduler.

Glow now support several backends, including single-core CPU/Interpreter and OpenCL. But as @stoklund mentioned in #1227 , the backend can be complicated. As for how to mapping a computation graph to some backend, we can have different scheduling schema. #1268 by @bertmaher is a kind of scheduling schema.

The topic is about different backends and different graph scheduling schema. I am interested in this topic, and want to know the long-term target of Glow at this topic. Thanks!

File: tests/unittests/graphSchedulerTest.cpp [hunk 1]
```diff
@@ -87,3 +89,47 @@ TEST(GraphScheduler, testMaxSizeLessThanResultSize) {
   EXPECT_LT(std::distance(schedule.begin(), sliceBigIt),
             std::distance(schedule.begin(), concatSmallIt));
 }
+
+TEST(GraphScheduler, testTopologicalSort) {
+  Module MD;
+  Context ctx;
+  auto *smallTensorA =
+      MD.createPlaceholder(ElemKind::FloatTy, {1, 4, 4}, "small_1", false);
+  ctx.allocate(smallTensorA);
```
Comment (nadavrot) [inline tests/unittests/graphSchedulerTest.cpp]: You don't need to allocate anything into the context here because you are not executing anything.

File: tests/unittests/graphSchedulerTest.cpp [hunk 2]
```diff
@@ -23,6 +23,8 @@
 
 #include "gtest/gtest.h"
 
+#include <iostream>
```
Comment (nadavrot) [inline tests/unittests/graphSchedulerTest.cpp]: no iostream please.

File: lib/IR/GraphScheduler.cpp [hunk 3]
```diff
@@ -170,6 +188,48 @@ void ChildMemSizeBasedScheduler::schedule() {
   scheduleNodes();
 }
 
+void TopologicalSortBasedScheduler::schedule() {
```
Comment (nadavrot) [inline lib/IR/GraphScheduler.cpp]: Comments please.

File: lib/IR/GraphScheduler.cpp [hunk 4]
```diff
@@ -24,15 +24,33 @@
...
 using llvm::dyn_cast;
 using llvm::isa;
 
+namespace {
+llvm::cl::OptionCategory graphSchedulerCat("Graph Scheduler Options");
+
+llvm::cl::opt<SchedulerKind> graphScheduler(
+    llvm::cl::desc("Scheduler to use:"),
+    llvm::cl::values(clEnumValN(SchedulerKind::ChildMemSizeBased,
+                                "child_mem_size_based",
```
Comment (nadavrot) [inline lib/IR/GraphScheduler.cpp]: I think that the LLVM convention is not to use underscore in flag names. For example "show-hidden" and not "show_hidden".

**Comment by nadavrot** [comment]
@QiJune The design doc makes sense to me. The problem is stated clearly and is very relevant for Glow. Concurrent execution is something that we should prepare to solve. I don't think that we are ready yes as we are still refactoring major parts of the compiler, such as the execution engine. I think that we should wait a few months before we start the design discussions of the graph partitioner. 

 I don't think that we have a need for topological sort right now but the changes are not too intrusive. I am okay with merging these changes if you clean things up. 

Quick question. If you change the scheduling policy are the tests still passing? Could you add a RANDOMIZE scheduling policy to shake things up? (random schedule should still be a legal and deterministic).

File: lib/IR/GraphScheduler.cpp [hunk 5]
```diff
@@ -170,6 +188,48 @@ void ChildMemSizeBasedScheduler::schedule() {
...
+      }
+    }
+  }
+}
+
+Scheduler *createScheduler(SchedulerKind schedulerKind, Function &G,
+                           NodesPtrList &scheduled) {
+  switch (schedulerKind) {
+  case SchedulerKind::ChildMemSizeBased:
+    return new ChildMemSizeBasedScheduler(G, scheduled);
```
Comment (rdzhabarov) [inline lib/IR/GraphScheduler.cpp]: can you move schedulers into its own files?

**Comment by nadavrot** [comment]
@QiJune I think that this PR is almost ready to go. Just a few small fixes and we'll be able to land it.

File: lib/IR/GraphScheduler.cpp (hunk 4 above)
Comment (QiJune) [reply lib/IR/GraphScheduler.cpp]: Done

File: lib/IR/GraphScheduler.cpp (hunk 3 above)
Comment (QiJune) [reply lib/IR/GraphScheduler.cpp]: Done

File: tests/unittests/graphSchedulerTest.cpp (hunk 2 above)
Comment (QiJune) [reply tests/unittests/graphSchedulerTest.cpp]: Done

File: tests/unittests/graphSchedulerTest.cpp (hunk 1 above)
Comment (QiJune) [reply tests/unittests/graphSchedulerTest.cpp]: Done

File: lib/IR/GraphScheduler.cpp (hunk 5 above)
Comment (QiJune) [reply lib/IR/GraphScheduler.cpp]: Not very clear about this, do you mean spilit GraphScheduler.cpp into two files, ChildMemSizeBasedScheduler.cpp and TopologicalSortBasedScheduler.cpp?

**Comment by QiJune** [comment]
@nadavrot Got it! Thanks for your detailed explanation. I will write some experimental codes in my own branch first.

**Comment by opti-mix** [comment]
@QiJune BTW, since you implement a topological sort graph scheduler, you may be interested in the `GraphPostOrderVisitor`, which essentially traverses the graph in a topological order. A schedule based on it would look like:

```
    GraphPostOrderVisitor gpov(G_);
    auto PostOrderNodes = gpov.getPostOrder();
    for (auto &N : PostOrderNodes) {
      Scheduled_.push_back(N);
    }
```

**Comment by opti-mix** [comment]
>>    Quick question. If you change the scheduling policy are the tests still passing? Could you add a RANDOMIZE scheduling policy to shake things up? (random schedule should still be a legal and deterministic).

> Not very clear about this. Do you mean testing the new scheduling policy for all the unittests?
I have tested the topological schedule in renset18 example, the result is the same with the default schedule.

We have a script called `run.sh` in `tests/images`. This script run a bunch of tests on real NN models. It would be nice to see if all of them can still run and produce correct results even with your new scheduler (or using randomized scheduler selection as proposed by @nadavrot)

File: lib/IR/GraphScheduler.cpp (hunk 5 above)
Comment (QiJune) [reply lib/IR/GraphScheduler.cpp]: So the file structure will be 

```
IR---
    ChildMemSizeBasedScheduler.cpp
    CMakeLists.txt
    GraphScheduler.cpp
    GraphScheduler.h
    Instrs.cpp
    IR.cpp
    IRBuilder.cpp
    IRGen.cpp
    IRUtils.cpp
    TopologicalSortBasedScheduler.cpp
```

Or do we need another directory `GraphScheduler`?

```
IR---
    CMakeLists.txt
    GraphScheduler.cpp
    GraphScheduler.h
    GraphScheduler---
         ChildMemSizeBasedScheduler.h
         ChildMemSizeBasedScheduler.cpp
         TopologicalSortBasedScheduler.h
         TopologicalSortBasedScheduler.cpp
    Instrs.cpp
    IR.cpp
    IRBuilder.cpp
    IRGen.cpp
    IRUtils.cpp
```

@rdzhabarov @nadavrot 
Which one do you prefer?

Comment (rdzhabarov) [reply lib/IR/GraphScheduler.cpp]: The same folder is fine, we always can add that if needed later.

**Comment by QiJune** [comment]
@nadavrot @opti-mix I run both these two schedulers with `run.sh`, here is the [result](https://gist.github.com/QiJune/e8a670172394544567de6531058d5664).

Most of the results are the same, there are some slight differences in the probability of some model.

- ChildMemSizeBasedScheduler
```
Model: googlenet_v1_slim/googlenet_v1_slim.onnx
File: tests/images/imagenet/zebra_340.png      Label-K1: 340 (probability: 0.9668)
Model: googlenet_v4_slim/googlenet_v4_slim.onnx
File: tests/images/imagenet_299/cat_281_299.png        Label-K1: 281 (probability: 0.6368)
Model: googlenet_v4_slim/googlenet_v4_slim.onnx
File: tests/images/imagenet_299/dog_207_299.png        Label-K1: 207 (probability: 0.7259)
Model: googlenet_v4_slim/googlenet_v4_slim.onnx
File: tests/images/imagenet_299/zebra_340_299.png      Label-K1: 340 (probability: 0.8901)
```

- TopologicalSortBasedScheduler
```
Model: googlenet_v1_slim/googlenet_v1_slim.onnx
 File: tests/images/imagenet/zebra_340.png      Label-K1: 340 (probability: 0.9668)
Model: googlenet_v4_slim/googlenet_v4_slim.onnx
 File: tests/images/imagenet_299/cat_281_299.png        Label-K1: 281 (probability: 0.6375)
Model: googlenet_v4_slim/googlenet_v4_slim.onnx
 File: tests/images/imagenet_299/dog_207_299.png        Label-K1: 207 (probability: 0.7267)
Model: googlenet_v4_slim/googlenet_v4_slim.onnx
 File: tests/images/imagenet_299/zebra_340_299.png      Label-K1: 340 (probability: 0.8909)
```


Besides, I also find some errors in both schedulers:

```
Model: densenet121
File: tests/images/imagenet/cat_285.pngimage-classifier: ../tools/loader/ImageClassifier.cpp:305: std::vector<std::pair<float, long unsigned int> > getTopKPairs(glow::Handle<float>): Assertion `H.dims().size() == 1 && "H must be a Handle of a 1d Tensor."' failed.
tests/images/run.sh: line 7:  1189 Aborted                 (core dumped) ./bin/
```

**Comment by QiJune** [comment]
@opti-mix Yes, we can also use `GraphPostOrderVisitor ` to get the topological sort result. There are many ways to do a topological sorting. 

But I find that `PlaceHolder` is included in this way.

For the graph defined in [graphScheduleTest](https://github.com/pytorch/glow/blob/master/tests/unittests/graphSchedulerTest.cpp#L53), the result is following:
```
Placeholder
name : "small_1"
output : float<1 x 4 x 4>
users : 1
trainable : 0

Placeholder
name : "small_2"
output : float<1 x 4 x 4>
users : 1
trainable : 0

Concat
name : concatSmall
Dim : 0
users : 1
Inputs0 : float<1 x 4 x 4>
Inputs1 : float<1 x 4 x 4>
Result : float<2 x 4 x 4>

Placeholder
name : "big"
output : float<100 x 4 x 4>
users : 1
trainable : 0

Transpose
name : transposeBig
Input : float<100 x 4 x 4>
Shuffle : [0, 2, 1]
users : 1
Result : float<100 x 4 x 4>

Slice
name : sliceBig
Input : float<100 x 4 x 4>
Start : [0, 0, 0]
users : 1
Result : float<1 x 4 x 4>

Concat
name : concat
Dim : 0
users : 0
Inputs0 : float<2 x 4 x 4>
Inputs1 : float<1 x 4 x 4>
Result : float<3 x 4 x 4>
```

In [ChildMemSizeBasedScheduler](https://github.com/pytorch/glow/blob/master/lib/IR/GraphScheduler.cpp#L187), it seems that `PlaceHolder` and `Constant` are exclueded.

**Comment by opti-mix** [comment]
@QiJune 

> Yes, we can also use GraphPostOrderVisitor to get the topological sort result. There are many ways to do a topological sorting.

> But I find that PlaceHolder is included in this way.

Yes, but you can easily skip any nodes you don't need in the schedule:

```c++
    GraphPostOrderVisitor gpov(G_);
    auto postOrderNodes = gpov.getPostOrder();
    for (auto &N : postOrderNodes) {
      // Skip any nodes we don't need in the schedule.
      if (isa<Placeholder>(N))
         continue;
      scheduled_.push_back(N);
    }
```

My main point was that we have an existing class `GraphPostOrderVisitor` to create a topological sort-based schedule. So, unless you need to have a more fancy, customizable class for producing the schedule, you could just use the existing code. But I've no strong opinion here.

**Comment by opti-mix** [comment]
@QiJune 

> Besides, I also find some errors in both schedulers:

>Model: densenet121
File: tests/images/imagenet/cat_285.pngimage-classifier: ../tools/loader/ImageClassifier.cpp:305: std::vector<std::pair<float, long unsigned int> > getTopKPairs(glow::Handle<float>): Assertion `H.dims().size() == 1 && "H must be a Handle of a 1d Tensor."' failed.
tests/images/run.sh: line 7:  1189 Aborted                 (core dumped) ./bin/

Hmm, this is interesting. Could you provide more context? What are the NN models in question? What is the exact command line?

**Comment by QiJune** [comment]
@opti-mix 

The command is 
```
./bin/image-classifier tests/images/imagenet/cat_285.png -image_mode=neg128to127 -m=squeezenet -model_input_name=data -cpu
```
I check the dims of output, it's a 3D tensor, with shape [1000, 1, 1].
squeezenet and densenet121 have this error. Other model is fine.

**Comment by rdzhabarov** [comment]
> I check the dims of output, it's a 3D tensor, with shape [1000, 1, 1].
squeezenet and densenet121 have this error. Other model is fine.

@SplitInfinity maybe we need to consider this case as well when we forcefully apply softmax

**Comment by SplitInfinity** [comment]
@rdzhabarov 

I already handled the case in which the net does not have a softmax layer and the output is not 1-dimensional: https://github.com/pytorch/glow/commit/55f72a81a3a2397d1017ca4a9631b8a4d425de04#diff-76c126707531603600e7f2612d2abd00R390

**Comment by opti-mix** [comment]
The stacktrace for the crash is:
```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00007fff58c19b86 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff58ccfc50 libsystem_pthread.dylib`pthread_kill + 285
    frame #2: 0x00007fff58b831c9 libsystem_c.dylib`abort + 127
    frame #3: 0x00007fff58b4b868 libsystem_c.dylib`__assert_rtn + 320
  * frame #4: 0x00000001000361a2 image-classifier`getTopKPairs(H=Handle<float> @ 0x00007ffeefbfe480) at ImageClassifier.cpp:305
    frame #5: 0x000000010002d661 image-classifier`processAndPrintResults(SMVarT=0x0000000103d14540, functionName=(Data = "squeezenet", Length = 10)) at ImageClassifier.cpp:371
    frame #6: 0x000000010002bf19 image-classifier`main(argc=6, argv=0x00007ffeefbff3d0) at ImageClassifier.cpp:433
    frame #7: 0x00007fff58adb085 libdyld.dylib`start + 1
```

So, it fails in `getTopKPairs` and it has nothing to do with the current PR.

@jfix71  Could you have a look at this? IIRC, you recently introduced this `getTopKPairs` funtionality.

File: lib/IR/TopologicalSortBasedScheduler.cpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  GraphPostOrderVisitor gpov(G_);
+  auto PostOrderNodes = gpov.getPostOrder();
+  for (auto &N : PostOrderNodes) {
+    if (isa<Storage>(N)) {
+      continue;
+    }
+    scheduled_.push_back(N);
+  }
+}
+} // namespace glow
```
Comment (opti-mix) [inline lib/IR/TopologicalSortBasedScheduler.cpp]: Add a new line after this line, please.

File: lib/IR/ChildMemSizeBasedScheduler.cpp [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+      orderChildNodesAndSchedule(&N);
+  }
+}
+
+void ChildMemSizeBasedScheduler::schedule() {
+  computeNodeResultsMemorySize();
+  computeNodeComputationMaxMemorySize();
+  scheduleNodes();
+}
+} // namespace glow
```
Comment (opti-mix) [inline lib/IR/ChildMemSizeBasedScheduler.cpp]: Add a new line after this line.

File: lib/IR/GraphScheduler.h [hunk 8]
```diff
@@ -21,6 +21,12 @@
 #include <unordered_map>
 
 namespace glow {
+
```
Comment (nadavrot) [inline lib/IR/GraphScheduler.h]: Doxygen:
/// Specifies the kind of graph scheduling to perform. 

/// This is a heuristics that tries to minimize memory usage.

/// Performs a standard topological search

Comment (QiJune) [reply lib/IR/GraphScheduler.h]: Done

File: lib/IR/ChildMemSizeBasedScheduler.cpp (hunk 7 above)
Comment (QiJune) [reply lib/IR/ChildMemSizeBasedScheduler.cpp]: Done

**Comment by nadavrot** [comment]
Nice job!
